### PR TITLE
Rails Capstone Project

### PR DESCRIPTION
**Pull request by rex-9**
### Description

The mobile web application where users can manage their budget: users have a list of transactions associated with a category so that the owner can see how much money they spent and on what. 

This Ruby on Rails application allows the user to:
- register and log in, so that the data is private to them.
- introduce new transactions associated with a category.
- see the money spent on each category.

#### Design
- Followed these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size, and weight.
  - Layout: composition and space between elements.

Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) is added to the project and gave appropriate credit to the author in the README file.

#### Interactions
- Splash screen
  - A simple page with the name of the app, and links to the signup and log-in pages.

- Sign up and login pages
  - The user is able to register in the app with full name, email, and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon, and the total amount of all the transactions that belong to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page, the user could see the total amount for the category (the sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user clicks (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

#### Testing requirements
- Created unit and integration tests for all the most important components of the RoR application.

#### Technical requirements

- Used Postgres as your database.
- Used devise for authentication.
- Validated all user input to make sure that anyone with bad intentions cannot compromise your app.
- Used a view template engine of your choice (.erb, .slim, .haml).
- The project is deployed and accessible online.
- The database schema reflects the following structure:

image

### Project documentation

Here is the video presentation of the "Transactions on Rails" project
https://www.loom.com/share/8f05f554375d4621aee04d7c430e3606

It is a video with a **maximum length of 5 minutes**. The content of the video includes:

- a description of the project.
- a demo of the project features.
- highlights of some interesting piece of code or something I built that I am very proud of.

### CODE REVIEW

File: spec/model/category_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  it 'Icon is required' do
+    subject.icon = nil
+    expect(subject).to_not be_valid
+  end
+
+  it 'Author is required' do
+    subject.author_id = nil
+    expect(subject).to_not be_valid
+  end
+end
```
Comment (descholar-ceo) [inline spec/model/category_spec.rb]: - [ ] You are supposed to have positive test cases as well

File: spec/model/purchase_spec.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  it 'Amount should be integer' do
+    subject.amount = '20'
+    expect(subject).to_not be_valid
+  end
+
+  it 'Author is required' do
+    subject.author_id = nil
+    expect(subject).to_not be_valid
+  end
+end
```
Comment (descholar-ceo) [inline spec/model/purchase_spec.rb]: - [ ] You are supposed to have positive cases as well
- [ ] And remember to test all models
Screen Shot 2022-11-16 at 06 38 49

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
Hi @rex-9 🤝 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
👍🏽 Implemented almost all requirements ✔️ 
👍🏽 The PR has linters and with no errors ✔️ 
👍🏽 You have a professional readme file ✔️ 
👍🏽 Your PR is professionally described ✔️ 

## Required Changes ♻️
_Check the comments under the review._

- [ ] The link to the live demo is broken, kindly fix it
- [ ] Remember to write tests for all models and for all views, currently you testing `purchases, and categories,` but you need to add the remaining ones as well
- [ ] Remember to record a video presentation of your project as well

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ShadyShawkat — APPROVED** [review]
## Hi @rex-9, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- All linters are passing ✔️ 
- All tests are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 
- Your readme looks professional ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**